### DISCUSSION TITLE
How do I force pagebreak in jupyter cell for typst output

### DISCUSSION THREAD

**Discussion by FranzForstmayr** [quarto-dev/quarto-cli#11164] [Q&A] [answered]
### Description

Following document produces typst error:
error: pagebreaks are not allowed inside of container

````qmd
---
title: |
    Report
format:
  typst:
    papersize: a4
    margin:
      x: 2cm
      y: 2cm
    keep-typ: true
    keep-md: true
execute:
  echo: false
---

```{python}
from IPython.display import Markdown
from IPython.display import display
```

```{python}
display(Markdown("## Test 1"))
display(Markdown("Some Text"))
display(Markdown("{{< pagebreak >}}"))
display(Markdown("## Test 2"))
display(Markdown("Some Text"))
```

````

The page break in the last cell is wrapped into a typst block, but no page breaks are allowed in typst blocks.
Is there a way to get page break from within a cell?

**Comment by mcanouil** [comment]
I don't think you need to use `display()`, so first I suggest you try without.
Then if it did not work, try adding `output: asis` as a code cell option.

**Comment by cderv** [comment]
You need `#| output: asis` here because you are using `display()` to get a output written for every lines. Without it, if you look at the `keep-md: true`, you'll see some output div as usual cell output results, which you don't want here as you aimed to produced raw markdown. 

If this is not used, then only the last Markdown will work. So you would need to only used one per cell

````markdown
---
title: |
    Report
format:
  typst:
    papersize: a4
    margin:
      x: 2cm
      y: 2cm
    keep-typ: true
    keep-md: true
execute:
  echo: false
---

```{python}
from IPython.display import Markdown
```

```{python}
Markdown(r"""## Test 1
Some Text
{{< pagebreak >}}
## Test 2
Some Text""")
```
````

Hope it helps.

**Comment by FranzForstmayr** [comment]
This helps indeed. However I need to have images on each page, but I'm unable to combine this approach together with images.
````qmd
---
title: Parameter Report
format:
  typst:
    papersize: a4
    fontsize: "11pt"
    margin:
      x: 2cm
      y: 2cm
    template-partials: 
      - "typst-template.typ"
    keep-typ: true
    keep-md: true
execute:
  echo: false
---


```{python}
from IPython.display import Markdown
from IPython.display import display
import matplotlib.pyplot as plt
import numpy as np

```


# Data

```{python}
#| output: asis
x = np.linspace(0,10)
y1 = np.sin(x)
y2 = np.cos(x)

Markdown("## Experiment 1")
fig, axes = plt.subplots(squeeze=False)
axes[0, 0].plot(x, y1)
fig

Markdown(r"""## Test 1
Some Text
{{< pagebreak >}}
## Test 2
Some Text""")

Markdown("## Experiment 2")
fig2, axes = plt.subplots(squeeze=False)
axes[0, 0].plot(x, y2)
fig2


```
````

**Reply by cderv** [reply]
You need to split cells. You are currently mixing content that should not be output as raw (your plot), and content that should be raw (Markdown strings). 

If you split in several cells, you should get what you want. Did you try ?

**Reply by FranzForstmayr** [reply]
Yes, this would be the best, but then it's not possible to execute these blocks n times?
The number of experiments is depending on the input data and for each experiment there's a bunch of text / tables and two plots necessary.

**Reply by cderv** [reply]
I believe this is really a conflict between your use case and how `{{< pagebreak >}}` shortcodes works... 
Maybe it is possible to use Lua filter to "clean" the div structure about the `{{< pagebreak >}}` part so that it does not end up into a output block. 

You need to have this `pagebreak` part as raw Markdown (so `output: asis`) and it is not possible to mixed content in a single cell. Limitation of what Quarto does and what Jupyter support. So unfortunately, I don't have easy solution in mind...

**Comment by FranzForstmayr** [comment]
After a few more tests I'm a little bit closer to the solution I think

````qmd
---
title: Parameter Report
format: 
  typst:
    papersize: a4
    fontsize: "11pt"
    margin:
      x: 2cm
      y: 2cm
    keep-typ: true
    keep-md: true
execute:
  echo: false
ipynb-shell-interactivity: none
---


```{python}
from IPython.display import Markdown
from IPython.display import display
import matplotlib.pyplot as plt
import numpy as np

```


# Data
Introduction Page....

```{python}
#| output: asis
for i in range(5):
  display(Markdown("""
{{< pagebreak >}}
"""))
  display(Markdown(f"## Experiment {i+1}"))
  x = np.linspace(0,i+3)
  y1 = np.sin(x)

  fig, axes = plt.subplots(squeeze=False)
  axes[0, 0].plot(x, y1)
  display(fig)
  fig, axes
  display(Markdown(
"""
### Result
this was really good

"""))
```
````

I still have the issue that all the plots arrive at the end of the block again?  I thought `ipynb-shell-interactivity: none` should supress this behaviour? 

Below is the intermediate md file:
````md
---
title: Parameter Report
format: 
  typst:
    papersize: a4
    fontsize: "11pt"
    margin:
      x: 2cm
      y: 2cm
    keep-typ: true
    keep-md: true
execute:
  echo: false
ipynb-shell-interactivity: none
---




# Data
Introduction Page....



{{< pagebreak >}}



## Experiment 1
![](test_files/figure-typst/cell-3-output-3.svg){}

### Result
this was really good





{{< pagebreak >}}



## Experiment 2
![](test_files/figure-typst/cell-3-output-7.svg){}

### Result
this was really good





{{< pagebreak >}}



## Experiment 3
![](test_files/figure-typst/cell-3-output-11.svg){}

### Result
this was really good





{{< pagebreak >}}



## Experiment 4
![](test_files/figure-typst/cell-3-output-15.svg){}

### Result
this was really good





{{< pagebreak >}}



## Experiment 5
![](test_files/figure-typst/cell-3-output-19.svg){}

### Result
this was really good



![](test_files/figure-typst/cell-3-output-21.svg){}
![](test_files/figure-typst/cell-3-output-22.svg){}
![](test_files/figure-typst/cell-3-output-23.svg){}
![](test_files/figure-typst/cell-3-output-24.svg){}
![](test_files/figure-typst/cell-3-output-25.svg){}



```

**Reply by mcanouil** [reply]
Why are you setting `none`? I believe what you want is `all` isn't it?

Note that as documented this option is for "top-level expression" which means content of loops are not controlled by this.

https://quarto.org/docs/computations/execution-options.html#jupyter-options

https://ipython.readthedocs.io/en/stable/config/options/terminal.html#configtrait-InteractiveShell.ast_node_interactivity

**Reply by FranzForstmayr** [reply]
The actual value of `ipynb-shell-interactivity` does not make any difference.
Basically I want the output of the explicit `display` calls, which works fine, without the implicit output of all images again.

In this case `test_files/figure-typst/cell-3-output-3.svg` is the same image as `test_files/figure-typst/cell-3-output-21.svg`, but only the first one is in the correct place.

**Reply by mcanouil** [reply]
> The actual value of ipynb-shell-interactivity does not make any difference.
> Basically I want the output of the explicit display calls, which works fine, without the implicit output of all images again.

As I said and as documented "top-level expression".

**Reply by FranzForstmayr** [reply, marked as answer]
Sorry I missunderstood... I thought only top-level expressions are collected and the rest is ignored.
However I think I found a working solution: `plt.ioff()`

````qmd
---
title: Parameter Report
format: 
  typst:
    papersize: a4
    fontsize: "11pt"
    margin:
      x: 2cm
      y: 2cm
    keep-typ: true
    keep-md: true
    keep-ipynb: true
execute:
  echo: false
ipynb-shell-interactivity: all
---


```{python}
#| output: false
from IPython.display import Markdown
from IPython.display import display
import matplotlib.pyplot as plt
import numpy as np
plt.ioff()

```


# Data
Introduction Page....

```{python}
#| output: asis
for i in range(5):
  display(Markdown("""
{{< pagebreak >}}
"""))
  display(Markdown(f"## Experiment {i+1}"))
  x = np.linspace(0,i+3)
  y1 = np.sin(x)

  fig, axes = plt.subplots(squeeze=False)
  axes[0, 0].plot(x, y1)
  display(fig)
  fig, axes
  display(Markdown(
"""
### Result
this was really good

"""))
```
````
